python/pyarrow/_flight.pyx

```diff
@@ -1655,70 +1655,83 @@ cdef CStatus _data_stream_next(void* self, CFlightPayload* payload) except *:
+                            "RecordBatch, or RecordBatchStreamReader objects, "
+                            "not {}.".format(type(result)))
+        # Don't loop around
         return CStatus_OK()
-    except FlightError as flight_error:
-        return (<FlightError> flight_error).to_status()
-
-    if isinstance(result, (list, tuple)):
-        result, metadata = result
-    else:
-        result, metadata = result, None
-
-    if isinstance(result, (Table, RecordBatchReader)):
-        if metadata:
-            raise ValueError("Can only return metadata alongside a "
-                             "RecordBatch.")
-        result = RecordBatchStream(result)
-
-    stream_schema = pyarrow_wrap_schema(stream.schema)
-    if isinstance(result, FlightDataStream):
-        if metadata:
-            raise ValueError("Can only return metadata alongside a "
-                             "RecordBatch.")
-        data_stream = unique_ptr[CFlightDataStream](
-            (<FlightDataStream> result).to_stream())
-        substream_schema = pyarrow_wrap_schema(data_stream.get().schema())
-        if substream_schema != stream_schema:
-            raise ValueError("Got a FlightDataStream whose schema does not "
-                             "match the declared schema of this "
-                             "GeneratorStream. "
-                             "Got: {}\nExpected: {}".format(substream_schema,
-                                                            stream_schema))
-        stream.current_stream.reset(
-            new CPyFlightDataStream(result, move(data_stream)))
-        return _data_stream_next(self, payload)
-    elif isinstance(result, RecordBatch):
-        batch = <RecordBatch> result
-        if batch.schema != stream_schema:
-            raise ValueError("Got a RecordBatch whose schema does not "
-                             "match the declared schema of this "
-                             "GeneratorStream. "
-                             "Got: {}\nExpected: {}".format(batch.schema,
-                                                            stream_schema))
-        check_flight_status(GetRecordBatchPayload(
-            deref(batch.batch),
-            stream.c_options,
-            &payload.ipc_message))
-        if metadata:
-            payload.app_metadata = pyarrow_unwrap_buffer(as_buffer(metadata))
-    else:
-        raise TypeError("GeneratorStream must be initialized with "
-                        "an iterator of FlightDataStream, Table, "
-                        "RecordBatch, or RecordBatchStreamReader objects, "
-                        "not {}.".format(type(result)))
-    return CStatus_OK()
+    # Ran out of attempts (the RPC handler kept yielding empty tables/readers)
+    raise RuntimeError("While getting next payload, ran out of attempts to "
+                       "get something to send "
+                       "(application server implementation error)")
```

Comment:
Why don't we simply send an empty batch here? It would sound more user-friendly then forbidding certain edge cases.

Reply:
I wanted to do that, but:
- An empty record batch reader would have the same issue, and is harder to detect. (Granted, it would be a rather weird application, but it's possible. We could just ignore this as an especially esoteric edge case.)
- In my interpretation, an empty table has no rows, while an empty batch has 0 rows - so it's not quite the same semantically.

That said I'm not opposed to just sending an empty batch and perhaps punting on the empty reader issue.

Reply:
What do you mean with an empty record batch reader? Would that be an empty generator in Python?

Reply:
An empty generator, or a reader for an empty file, or something of that sort, yes.

Reply:
Though to be clear just in case, this function is adapting a generator of (batches, tables, readers, etc.) into a generator of just batches so I mean the inner table/reader/etc. here.

Reply:
Fair enough. Perhaps the test should just use such an empty generator, then? I was surprised to see you needed an explicit empty table to trigger the issue.

Reply:
I misspoke: we don't allow a subgenerator, only a sub-table or sub-reader, so a sub-table was the easiest way to trigger this. (And normally, an empty table or reader is completely fine and will just be skipped: this is guarding against the rather esoteric issue of an infinite stream of empty tables…)